README.md
# sind


A semi-magical list-friendly selection dialog for Bash 4+ with reasonable defaults. Features single and multiple choice modes, and can display the option list on a single line. Developed on Ubuntu 18+ through [UserLAnd on Android](https://play.google.com/store/apps/details?id=tech.ula).


![CI](https://github.com/l3laze/sind/workflows/CI/badge.svg) [![Codacy Badge](https://app.codacy.com/project/badge/Grade/3212c5503ee94a42adb04cd730304493)](https://www.codacy.com/manual/l3laze/sind?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=l3laze/sind&amp;utm_campaign=Badge_Grade) [![codecov](https://codecov.io/gh/l3laze/sind/branch/master/graph/badge.svg)](https://codecov.io/gh/l3laze/sind)


[![Generic badge](https://img.shields.io/badge/Made%20with-Bash-black.svg)](https://shields.io/) [![Only 8 Kb](https://badge-size.herokuapp.com/l3laze/sind/master/sind.sh)](https://github.com/l3laze/sind/blob/master/sind.sh)


----


<details><summary><b>Table of contents</b></summary>

  * [Demo](#Demo)
  * [Usage](#Usage)
    - [Options](#Options)
    - [Example](#Example)
    - [Output](#Output)
  * [Installation](#Installation)
    - [Automatic](#Automatic)
    - [Manual](#Manual)
  * [Credits](#Credits)
  * [Contributing](#Contributing)
  * [Unlicense](#Unlicense)
</details>


----


## Demo


![sind.sh demo made using asciinema + Termux on Android](https://user-images.githubusercontent.com/18404758/90070297-3f764f00-dcb9-11ea-96d3-2f399d782205.gif)


----


## Usage


`sind [options...]`


### Options


| Option | What? | Default |
| --- | --- | --- |
| -c, --cancel | Add cancel to end of options if it doesn't exist. | - |
| -h, --help | Display this message. | - |
| -l, --line | Single-line list mode. | - |
| -m, --multiple | Multiple choice mode. | - |
| -o, --options | List of options, space-separated (requires at least one arg). | Yes No |
| -t, --title | Prompt printed above list. | Choose one/Choose some |
| -v, --version | Print version. | - |
| --marker | Character used to mark selected options in multiple choice mode. | > |


### Example


```sh
echo "$(sind -t Title -o Okay -c)"
```


### Output:


```sh
Title
(up|j, down|k, enter: choose)
----------------------------------
Okay
Cancel
----------------------------------
Okay
```


----


## Installation


### Automatic


```sh
# Download fully and then run:
sh -c "$(curl -sSL https://raw.githubusercontent.com/l3laze/sind/install.sh)"


# Run lines while downloading:
curl -sSL https://raw.githubusercontent.com/l3laze/sind/install.sh | bash
```


This will install to /usr/local/bin.


### Manual


* Download `sind.sh`, optionally rename it to `sind`.
* Move it to a directory that's available from your `$PATH`, or to the project that you're working on.
* Make it executable with `chmod u+x ./sind.sh` (or just sind, if you changed it).


For a single user it can be installed to `$HOME/bin`  (though it may not exist by default). For a system-wide install `/usr/local/bin` is best, based on the [Linux filesystem hierarchy](https://linux.die.net/man/7/hier) and related answers on [StackExchange](https://unix.stackexchange.com/questions/8656/usr-bin-vs-usr-local-bin-on-linux).


----


## Credits


- [Alexander Klimetschek's Stack Exchange answer](https://unix.stackexchange.com/a/415155/310780)
- [mtfurlan's vi keybinding patch](https://github.com/l3laze/sind/issues/1)


----


## Contributing


Please ensure your pull request adheres to the following guidelines:


- **Lint with [shellcheck](https://www.shellcheck.net/)**. This is done on push, but doing so while developing helps minimize senseless builds which are a waste of resources.
- **[Open an issue](https://github.com/l3laze/sind/issues) before creating any new features**, for discussion.
- Bug fixes & POSIX-friendly changes are more than welcome!
- Check your spelling and grammar (if applicable).
- Try to use a text editor that will remove trailing whitespace, or configure git to do so.
- Run `visual-test.sh` to confirm visual output of tests is proper.


----


## Unlicense


<details><summary><b>This project is distributed under The Unlicense:</b></summary>


<br/>


This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>

</details>


<br />


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v4.0.0] - {date}
### Added
- CHANGELOG.md -- \*hint hint\*.
- kcov for coverage.
- visual-test.sh for visually checking results.
- Multiple choice feature.
- Single line option-list feature.
- -c|--cancel option, to add cancel if it's not provided.

### Changed
- Cancel is no longer a default option.
- Title is no longer required.
- test.sh now measures second.millisecond runtime instead of seconds.
- Using VERSION file rather than a hard-coded variable.
- Refractored internal structure to help support here-string input for automation, to ease testing.

### Fixed
- Moved to GitHub CI from Travis-CI, where it just works and coverage can be properly measured.

### Removed
- No more arguments to install.sh.
- Removed `set -euo pipefail` throughout.


----


## [v3.0.0] - 2020/6/17
### Added
- Travis-CI (for shellcheck).
- CLI args as commands: line|list=mode, version, help.
- Installation script install.sh.


### Changed
- Refactored to use ANSI escape sequences instead of clearing and drawing the screen repeatedly.


----


## [v2.0.0] - 2020/6/1
### Changed
- Output selection text rather than index.


----


## [v1.1.0] - 2020/6/1
### Added
- Multiline title support.
- Use vim key binding aliases -- j/k as up/down.
- Output selection index.


test.sh
#!/usr/bin/env bash

set -euo pipefail

if command -v "shellcheck" > /dev/null 2>&1; then
  echo "shellcheck"
  shellcheck ./*.sh && printf "  ✓ sind.sh\n  ✓ install.sh\n  ✓ test.sh\n  ✓ visual-test.sh\n" || exit 64
else
  echo >&2 "Need shellcheck installed for linting." # LCOV_EXCL_LINE
fi

run () {
  local timer
  local label
  local actual
  local expected
  local passed=0
  local total=0

  timer="$(date +%s%3N)"

  test () {
    label="$1"
    actual="$2"
    expected="$3"
    ((total += 1))

    if [[ "$actual" == *"$expected"* ]]; then
      printf "  ✓ %s\n" "$label"
      ((passed += 1))
    else
      printf "  × %s\n%s != %s\n" "$label" "$expected" "$actual" # LCOV_EXCL_LINE
    fi
  }

  echo "install.sh"

  test "Installs latest release automatically from GitHub" "$(./install.sh 2>&1)" "Success"


  echo "sind.sh"


  # SHOULD PASS

  test "Prints usage" "$(./sind.sh -h 2>&1)" "Usage"

  test "Prints version" "$(./sind.sh -v 2>&1)" "v"

  test "Uses default title" "$(./sind.sh <<< $'\n' 2>&1)" "Choose one"

  test "Takes a title" "$(./sind.sh -t "title goes here" <<< $'\n' 2>&1)" "title goes here"

  test "Takes an option" "$(./sind.sh -o Okay <<< $'\n' 2>/dev/null)" "Okay"

  test "Handles here-string input" "$(./sind.sh <<< $'\e[2A\n' 2>/dev/null)" "Yes"

  test "Multiple choice" "$(./sind.sh -m <<< $' \e[B \n' 2>/dev/null)" $'Yes\nNo'

  test "Press any key to continue" "$(./sind.sh -m <<< $'\n   \e[A \n' 2>/dev/null)" "No"

  test "Removes de-selected choices" "$(./sind.sh -m <<< $'  \e[B \n' 2>/dev/null)" "No"

  test "Line mode" "$(./sind.sh -l <<< $'\e[B\n' 2>/dev/null)" "No"

  test "Combo line mode + multiple choice" "$(./sind.sh -l -m <<< $' \e[B \n' 2>/dev/null)" $'Yes\nNo'

  test "Arg -c|--cancel adds cancel if not provided" "$(./sind.sh -c <<< $'\e[A\n' 2>/dev/null)" "Cancel"

  test "Arg --marker changes selection mark" "$(./sind.sh -m --marker + <<< $' \e[B\n' 2>&1)" "+Yes"

  test "Splits long title from directions" "$(./sind.sh -t 'This should be a sufficiently long title to cause it to have a newline between itself and the directions' <<< $'\n' 2>&1)" $'This should be a sufficiently long title to cause it to have a newline between itself and the directions\n('


  # SHOULD FAIL

  test "Fails with invalid options" "$(./sind.sh -x 2>&1)" "Error: Unknown option: -x"

  test "Fails with -t|--title and no arg" "$(./sind.sh -t 2>&1)" "Error: Title must be at least one character."

  test "Fails if arg to -t|--title is empty" "$(./sind.sh -t '' 2>&1)" "Error: Empty title is not allowed."

  test "Fails with -o|--options and no args" "$(./sind.sh -o 2>&1)" "Error: Options must be at least one character."

  test "Fails with empty arg to -o|--options" "$(./sind.sh -o '' 2>&1)" "Error: Empty options are not allowed."

  test "Fails with --marker and no args" "$(./sind.sh --marker 2>&1)" "Error: Marker must be one character."

  test "Fails if arg to --marker is > 1 character" "$(./sind.sh --marker 69 2>&1)" "Error: Marker can't be more than one character."

  test "Fails if arg to --marker is empty" "$(./sind.sh --marker '' 2>&1)" "Error: Marker can't be empty."

  echo -e "\n$passed/$total passed"

  timer="$(($(date +%s%3N) - timer))"

  printf "Finished in %s.%s seconds\n" "$((timer / 1000))" "$((timer % 1000))"

  if [[ "$passed" -lt "$total" ]]; then
    exit 64 # LCOV_EXCL_LINE
  fi
}

run


tests/test.sh
#!/usr/bin/env bash
#
# Test script for sind

set -euo pipefail

# shellcheck disable=SC1091
source ./sind.sh

# From https://stackoverflow.com/a/42762743
hr () {
  printf '\n%*s\n' "${COLUMNS:-$(tput cols)}" '' | tr ' ' -
}

if [[ "${1:-}x" == "x" ]]; then
  echo >&2 "No test specified." && exit 1
elif [[ ! "$1" =~ [[:digit:]]+ ]]; then
  echo >&2 "Not a number: $1" && exit 1
fi

case "$1" in
  1)
    userChoice=$(sind "line" "Choose one...\nOf these." "  Yes   " "   No   " " Batman " " Cancel ")
    hr
    printf "Selected %s\n" "$userChoice"
  ;;

  2)
    userChoice=$(sind "list" "Choose one...\nOf these." "  Yes   " "   No   " " Batman " " Cancel ")
    hr
    printf "Selected %s\n" "$userChoice"
  ;;

  3)
    userChoice=$(sind "list" "Choose one...\nThis is a big title.\nMany lines.\nMuch words.\nVery wow." "  Yes   " "   No   " " Batman " " Cancel ")
    hr
    printf "Selected %s\n" "$userChoice"
  ;;

  4)
    userChoice=$(sind "version" "Choose one...\nThis is the title." "One" "Two")
    # Should print version.
  ;;

  5)
    userChoice=$(sind "help" "Choose one...\nThis is the title." "One" "Two")
    # Should print usage.
  ;;

  6)
    userChoice=$(sind "Title." "  Yes   " "   No   " " Batman " " Cancel ")
    # Should error: unknown command.
  ;;

  7)
    userChoice=$(sind "line" "Choose one...\nThis is the title.")
    # Should error: not enough options.
  ;;

  8)
    userChoice=$(sind "line" "Choose one..." "  Yes   " "   No   " " Batman " " Cancel ")
    printf "Selected %s\n" "$userChoice"
  ;;

  *)
    echo "Invalid test number: $1"
esac

visual-test.sh
#!/usr/bin/env bash

fail () {
  echo -e >&2 "$1"
  # shellcheck disable=SC2086
  exit $2
}

run () {
  local timer

  timer="$(date +%s%3N)"

  echo -e "#\n# ******** PASSING TESTS ********\n#"

  printf "\n ===> %s <===\n" "Installs latest release automatically from GitHub"
  ./install.sh 2>&1 || fail "Did not auto install"

  printf "\n ===> %s <===\n" "Prints usage" 
  ./sind.sh -h 2>&1 || fail "Did not print usage"

  printf "\n ===> %s <===\n" "Prints version"
  ./sind.sh -v 2>&1 || fail "Did not print version"

  printf "\n ===> %s <===\n" "Uses defaults"
  ./sind.sh <<< $'\n' 2>&1 || fail "Did not use defaults"

  printf "\n ===> %s <===\n" "Takes a title arg"
  ./sind.sh -t "title goes here" <<< $'\n' 2>&1 || fail "Did not handle title arg"

  printf "\n ===> %s <===\n" "Takes an option arg"
  ./sind.sh -o Okay <<< $'\n' || fail "Did not handle option arg"

  printf "\n ===> %s <===\n" "Takes here-string input"
  ./sind.sh <<< $'\e[2A\n' || fail "Did not handle here-string input"

  printf "\n ===> %s <===\n" "Multiple-choice"
  ./sind.sh -m <<< $' \e[B\n \n' || fail "Did not handle multiple-choice"

  printf "\n ===> %s <===\n" "Press any key to continue"
  ./sind.sh -m <<< $'\n \e[A \n' || fail "Did not handle press any key to continue"

  printf "\n ===> %s <===\n" "Removes de-selected choices"
  ./sind.sh -m <<< $'  \e[B \n' || fail "Did not remove de-selected choices"

  printf "\n ===> %s <===\n" "Line mode"
  ./sind.sh -l <<< $'\e[B\n' || fail "Did not handle line mode"

  printf "\n ===> %s <===\n" "Combo multiple-choice + line mode"
  ./sind.sh -l -m <<< $' \e[B \n' || fail "Did not handle combo multiple-choice + line mode"

  printf "\n ===> %s <===\n" "Adds cancel"
  ./sind.sh -c <<< $'\e[A\n' || fail "Did not handle adding cancel"

  printf "\n ===> %s <===\n" "Custom selection marker"
  ./sind.sh -m --marker + <<< $' \e[B\n' || fail "Did not handle selection marker replacement"

  printf "\n ===> %s <===\n" "Splits title from directions"
  ./sind.sh -t 'This should be a sufficiently long title to cause it to have a newline between itself and the directions' <<< $'\n' 2>&1 || fail "Did not handle splitting title and directions"

  #
  # FAILURES
  #

  echo -e "\n#\n# ******** FAILING TESTS ********\n#"

  head -1 <<< "$(./sind.sh -x 2>&1 && fail 'Did not fail with invalid option')"
  ./sind.sh -t 2>&1 && fail "Did not fail with -t and no arg"
  ./sind.sh -t '' 2>&1 && fail "Did not fail with -t and empty arg"
  ./sind.sh -o 2>&1 && fail "Did not fail with -o and no args"
  ./sind.sh -o '' 2>&1 && fail "Did not fail with -o and empty arg"
  ./sind.sh --marker 2>&1 && fail "Did not fail with --marker and no arg"
  ./sind.sh --marker 69 2>&1 && fail "Did not fail with --marker and 2+ character arg"
  ./sind.sh --marker '' 2>&1 && fail "Did not fail with --marker and empty arg"

  timer="$(($(date +%s%3N) - timer))"

  printf "Finished in %s.%s seconds\n" "$((timer / 1000))" "$((timer % 1000))"
}

run

install.sh
#!/usr/bin/env bash

# Originally based on install.sh from NVM
# https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.0/install.sh

{
  user_has () {
    type "$1" >/dev/null 2>&1
  }

  do_download () {
    if user_has "curl"; then
      curl -s -q --compressed "$@"
    elif user_has "wget"; then # LCOV_EXCL_LINE
      # Emulate curl with wget
      ARGS=$(echo "$*" | command sed -e 's/--compressed //' -e 's/-o /-O /' -e 's/-C - /-c /') # LCOV_EXCL_LINE
      # shellcheck disable=SC2086
      eval wget $ARGS # LCOV_EXCL_LINE
    fi
  }

  function install () {
    local version

    if ! user_has "curl" && ! user_has "wget"; then # LCOV_EXCL_LINE
      echo >&2 "You must have curl or wget to use this install script." # LCOV_EXCL_LINE
      exit 65 # LCOV_EXCL_LINE
    else
      version="$(do_download -s https://api.github.com/repos/l3laze/sind/releases/latest)";
      version="${version/*\"tag_name\": \"/}"
      version="${version/\"*/}"
      echo >&2 "Installing latest sind from GitHub @ $version for current user."

      do_download -o "/usr/local/bin/sind" "https://raw.githubusercontent.com/l3laze/sind/${version}/sind.sh" || {
        echo >&2 "Failed to download sind.sh as /usr/local/bin/sind" # LCOV_EXCL_LINE
        exit 68 # LCOV_EXCL_LINE
      }
    fi

    chmod u+x "/usr/local/bin/sind" || { # LCOV_EXCL_LINE
      echo >&2 "Failed to chmod /usr/local/bin/sind" # LCOV_EXCL_LINE
      exit 69 # LCOV_EXCL_LINE
    }

    # shellcheck source=/dev/null
    source "$HOME/.profile"

    if ! command -v sind >/dev/null 2>&1; then # LCOV_EXCL_LINE
      echo >&2 "Installation failed!" # LCOV_EXCL_LINE
      exit 70 # LCOV_EXCL_LINE
    else
      echo >&2 "Success."
    fi
  }

  install "$@"
}


sind.sh
#!/usr/bin/env bash

set -euo pipefail

# This script is based on code by Alexander Klimetschek at
# https://unix.stackexchange.com/a/415155/310780

#
# Follows the suggested exit code range of 0 & 64-113
# from http://www.tldp.org/LDP/abs/html/exitcodes.html
#
# | Exit Code | Meaning |
# |    ---    |   ---   |
# |     0     |  No errors. |
# |    64     |  Unknown option. |
# |    65     |  Invalid arg for option. |
# |    66     |  Abnormal exit (ctrl + c, etc) |
# |    ---    |  --- |
#

{
  key_input () {
    local key
    IFS=; read -rsN 1
    key="$REPLY"
      if [[ "$key" =~ ^[A-Za-z0-9]$ ]]; then printf "%s" "$key";
    elif [[ "$key" == $'\n' ]]; then printf "enter";
    elif [[ "$key" == $' ' ]]; then printf "space";
    elif [[ "$key" == $'\e' ]]; then
      IFS=; read -rsN 2 -t 0.01
        if [[ "$REPLY" == "[A" ]]; then printf "up";
      elif [[ "$REPLY" == "[B" ]]; then printf "down";
      fi
    fi
    key=;
  }

  cursor_on () {
    printf >&2 "\e[?25h"
  }

  cursor_off () {
    printf >&2 "\e[?25l"
  }

  hr () {
    printf -v horule '%*s' "${COLUMNS:-$(tput cols)}" ''
    printf >&2 "%s" "${horule// /-}"
  }

  print_selected () {
    printf >&2 "%s" $'\e[7m'"$1"$'\e[27m'
  }

  sind () {
    local version="v4.0.1"
    local opts
    local selected=0
    local title
    local directions
    local choices=","
    local multiple=1
    local add_cancel=1
    local index=0
    local size=0
    local sel_mark=">"
    local usage
    local name="${0#\.\/}"

    opts=()
    name="${name%.sh}"
    usage="$name $version\n\nA semi-magical list-friendly selection dialog for Bash 4+ with reasonable defaults. Features single and multiple choice modes, and can display the option list on a single line.\n\nUsage $0 [options...]\n\nWhere options are:\n\n-c|--cancel\n     Add cancel to end of options if it doesn't exist\n-h|--help\n    Display this message\n-l|--line\n    Single-line list mode\n-m|--multiple\n    Multiple choice mode\n-o|--options = Yes No\n    Space-separated options (requires at least one arg)\n-t|--title = Choose one/some (requires one arg)\n    Prompt printed above list\n-v|--version\n    Print version\n--marker = >\n    Character used to mark selected options in multiple choice"

    cleanup () {
      printf >&2 "\e[%sB\n" "${#opts[@]}"
      [[ "$-" == *"i"* ]] && stty echo
      cursor_on
      # shellcheck disable=SC2086
      exit $1
    }

    trap "cleanup 66" HUP INT QUIT ABRT
    cursor_off
    [[ "$-" == *"i"* ]] && stty -echo

    while [[ "$#" -gt 0 ]]; do
      case "${1:-}" in
        -c|--cancel)
          shift
          if [[ "${opts[*],,}" != *"cancel"* ]]; then
            add_cancel=0
          fi
        ;;
        -h|--help)
          echo -en >&2 "$usage"
          cleanup 0
        ;;
        -l|--line)
          size=1
          shift
        ;;
        -m|--multiple)
          multiple=0
          shift
        ;;
        --marker)
          if [[ "$#" -lt 2 ]]; then
            printf >&2 "Error: Marker must be one character."
            cleanup 65
          fi
          shift
          sel_mark="$1"
          shift
          if [[ "${#sel_mark}" -lt 1 ]]; then
            printf >&2 "Error: Marker can't be empty."
            cleanup 65
          elif [[ "${#sel_mark}" -gt 1 ]]; then
            printf >&2 "Error: Marker can't be more than one character."
            cleanup 65
          fi
        ;;
        -o|--options)
          if [[ "$#" -lt 2 ]]; then
            printf >&2 "Error: Options must be at least one character."
            cleanup 65
          fi
          shift
          while [[ "$#" -gt 0 && ! "$1" =~ ^--? ]]; do
            if [[ "$1" == "" ]]; then
              printf >&2 "Error: Empty options are not allowed."
              cleanup 65
            fi
            opts+=("$1")
            shift
          done
        ;;
        -t|--title)
          if [[ "$#" -lt 2 ]]; then
            printf >&2 "Error: Title must be at least one character."
            cleanup 65
          fi
          shift
          title="$1"
          shift
          if [[ "${#title}" -lt 1 ]]; then
            printf >&2 "Error: Empty title is not allowed."
            cleanup 65
          fi
        ;;
        -v|--version)
          printf >&2 "%s" "$version"
          cleanup 0
        ;;
        *)
          printf >&2 "Error: Unknown option: %s." "$1"
          cleanup 64
        ;;
      esac
    done

    if [[ -z "${title:-}" ]]; then
      if [[ "$multiple" -eq 1 ]]; then
        title="Choose one"
      else
        title="Choose some"
      fi
    fi

    if [[ "${#opts[@]}" -eq 0 ]]; then
      opts=(Yes No)
    fi

    if [[ "$add_cancel" -eq 0 ]]; then
      opts+=(Cancel)
    fi

    if [[ "$multiple" -eq 1 ]]; then
      directions="(↑/j or ↓/|k, enter: choose)"
    else
      directions="(↑|j or ↓|k, space: select, enter: done)"
    fi

    if [[ "$((${#title} + ${#directions} + 1))" -lt "${COLUMNS:-$(tput cols)}" ]]; then
      echo -e >&2 "$title $directions" # LCOV_EXCL_LINE
    else
      echo -e >&2 "$title\n$directions"
    fi

    hr

    [[ "$size" -eq 1 ]] && printf "\n"

    while true; do
      if [[ "$size" -eq "1" ]]; then
        printf >&2 "\e[2K\e[1000D"

        if [[ ("$multiple" -eq 0 && "$choices" == *",${opts[$selected]},"*) ]]; then
          printf >&2 "%s" "$sel_mark"
        elif [[ "$multiple" -eq 0 ]]; then
          printf >&2 " "
        fi

        print_selected >&2 "${opts[$selected]}"
      else 
        for index in $(seq 0 "$((${#opts[@]} - 1))"); do
          printf >&2 "\n"

          if [[ ("$multiple" -eq "0" && "$choices" == *",${opts[$index]},"*) ]]; then
            printf >&2 "%s" "$sel_mark"
          elif [[ "$multiple" -eq 0 ]]; then
            printf >&2 " "
          fi

          if [[ "$index" == "$selected" ]]; then
            print_selected >&2 "${opts[$((index))]}"
          else
            printf "%s" "${opts[$((index))]}"
          fi

          index="((index + 1))"
        done

        printf >&2 "\e[%sA" "${#opts[@]}"
      fi

      case "$(key_input 2>/dev/null)" in # LCOV_EXCL_LINE
        'up'|'j')
          selected=$((selected - 1))

          if [ "$selected" -lt 0 ]; then selected=$(("${#opts[@]}" - 1)); fi
        ;;
        'down'|'k')
          selected=$((selected + 1))

          if [ "$selected" -gt $(("${#opts[@]}" - 1)) ]; then selected=0; fi
        ;;
        'space')
          if [[ "$multiple" -eq 0 ]]; then
            if [[ "$choices" == *",${opts[$selected]},"* ]]; then
              choices="${choices/,${opts[$selected]},/,}"
            else
              choices="${choices}${opts[selected]},"
            fi
          fi
        ;;
        'enter')
          if [[ "$size" -eq 0 ]]; then
            printf >&2 "\e[%sB\n" "${#opts[@]}"
          else
            printf >&2 "\n"
          fi

          hr

          if [[ "$multiple" -ne 0 ]]; then
            printf "%s" "${opts[$selected]}"
          else
            IFS=',' read -ra choice_array <<< "${choices#,}"
            if [[ "${#choice_array[@]}" -eq 0 ]]; then
              >&2 read -rsn 1 -p "Make a choice (press any key to continue)"

              printf >&2 "\e[1000D\e[2A\e[J"

              if [[ "$size" -ne 1 ]]; then
                printf >&2 "\e[%sA" "$((${#opts[@]}))"
              fi

              continue
            else
              for ((index = 0; index < "${#choice_array[@]}"; index++)); do
                printf "\n"
                printf "%s" "${choice_array[$index]}"
              done
            fi
          fi

          cleanup 0
        ;;
      esac
    done
  }

  sind "$@"
}
